### ISSUE TITLE
8-Deep learning with Tensorflow and Keras

### ISSUE THREAD

**Issue by K0nkere** [K0nkere/ml-bookcamp#5]
/

**Comment by K0nkere** [comment]
### Model creation on the base of pre-trained CNN

**Comment by K0nkere** [comment]
### Final model creation fuction
def make_model(learning_rate=0.01):
#     train_gen = ImageDataGenerator(preprocessing_function=preprocess_input)

#     train_ds = train_gen.flow_from_directory(
#         "./clothing-dataset-small/train/",
#         target_size=(150, 150),
#         batch_size=32,
#         shuffle=True
#         )
    
#     val_gen = ImageDataGenerator(preprocessing_function=preprocess_input)

#     val_ds = val_gen.flow_from_directory(
#         "./clothing-dataset-small/validation/",
#         target_size=(150, 150),
#         batch_size=32, 
#         shuffle=False
#         )
    
    base_model = Xception(
        weights="imagenet",
        input_shape=(150, 150, 3),
        include_top=False,
        )

    base_model.trainable = False
    
    #######################################################################
    
    inputs = keras.Input(shape=(150, 150, 3))
    base = base_model(inputs, training=False)
    
    vector = keras.layers.GlobalAveragePooling2D()(base)
    outputs = keras.layers.Dense(10)(vector)

    model = keras.Model(inputs, outputs)
    
    #######################################################################

    optimizer = keras.optimizers.Adam(learning_rate=learning_rate)

    loss = keras.losses.CategoricalCrossentropy(
                                from_logits=True
                    )
    model.compile(
        optimizer=optimizer,
        loss=loss,
        metrics=['accuracy']
        )
    
    return model

**Comment by K0nkere** [comment]
### Checkpointing
First of all its need to define checkpoint template
```
checkpoint = keras.callbacks.ModelCheckpoint(
        "xception_v1_{epoch:02d}_{val_accuracy:.3f}.h5",
        save_best_only=True,
        monitor="val_accuracy", # what kind of metric we want to monitor in order to save
        mode="max", # as soon as we want to maximize monitoring metric / if we want to minimixe rmse it will be mode="min" 
    )
```
and run training process with our previously tuned learning_rate
```
best_lr = 0.001
model = make_model(learning_rate=best_lr)
    
history = model.fit(
    train_ds,
    epochs=10,
    validation_data=val_ds,
    callbacks=[checkpoint]
    )
```

**Comment by K0nkere** [comment]
### Adding more Dense layers

**Comment by K0nkere** [comment]
### Tuning size of inner layer
(or learning rates)
```
best_lr = 0.001
inner_sizes = [10, 100, 1000]
scores = {}
    
for size_inner in inner_sizes:
    
    model = make_model(learning_rate=best_lr, size_inner=size_inner)
    history = model.fit(train_ds, epochs=10, validation_data=val_ds)
    scores[lr] = history
```

**Comment by K0nkere** [comment]
### Monitoring GPU utilization
in terminal while training
`monitor nvidia-smi`
